openc3-cosmos-cmd-tlm-api/app/controllers/storage_controller.rb

```diff
@@ -34,67 +34,151 @@ def buckets
     render :json => buckets, :status => 200
   end
 
+  def volumes
+    # ENV.map returns a big array of mostly nils which is why we compact
+    # The non-nil are MatchData objects due to the regex match
+    matches = ENV.map { |key, value| key.match(/^OPENC3_(.+)_VOLUME$/) }.compact
+    # MatchData [0] is the full text, [1] is the captured group
+    # downcase to make it look nicer, BucketExplorer.vue calls toUpperCase on the API requests
+    volumes = matches.map { |match| match[1].downcase }.sort
+    # Add a slash prefix to identify volumes separately from buckets
+    volumes.map! {|volume| "/#{volume}" }
+    render :json => volumes, :status => 200
+  end
+
   def files
     return unless authorization('system')
-    bucket = OpenC3::Bucket.getClient()
-    bucket_name = ENV[params[:bucket]] # Get the actual bucket name
-    path = params[:path]
-    path = '/' if path.nil? || path.empty?
-    # if user wants metadata returned
-    metadata = params[:metadata].present? ? true : false
-    results = bucket.list_files(bucket: bucket_name, path: path, metadata: metadata)
+    root = ENV[params[:root]] # Get the actual bucket / volume name
+    raise "Unknown bucket / volume #{params[:root]}" unless root
+    results = []
+    if params[:root].include?('_BUCKET')
+      bucket = OpenC3::Bucket.getClient()
+      path = params[:path].gsub('.', '')
+      path = '/' if path.nil? || path.empty?
+      # if user wants metadata returned
+      metadata = params[:metadata].present? ? true : false
+      results = bucket.list_files(bucket: root, path: path, metadata: metadata)
+    elsif params[:root].include?('_VOLUME')
+      dirs = []
+      files = []
+      list = Dir["/#{root}/#{params[:path].gsub('.', '')}/*"]
```

Comment:
Didn't think about that. I was trying to avoid going back with `../` or whatever. It's already an absolute path with the leading `/`. Trying to follow the recommendations here: https://codeql.github.com/codeql-query-help/java/java-path-injection/

Reply:
I guess its ok, just might burn us at some point in the future.